File: cliente-python/cliente.py
import os
import sys

from api_servico import ApiServico


def cabecalho(titulo):
    print()
    print(f"=== {titulo} ===")


def descrever_aluno(a):
    return f"{a['tipo']}[id={a['id']}, nome={a['nome']}, matricula={a['matricula']}]"


def main():
    base_url = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("SERVIDOR_URL", "http://localhost:8080")
    api = ApiServico(base_url)
    print(f"[cliente-python] conectado em {base_url}")

    try:
        cabecalho('1) criar_departamento("DEMA - Matematica")')
        d_ema = api.criar_departamento("DEMA - Matematica")
        print(f"    -> retorno: {d_ema['id']} ({d_ema['nome']})")

        cabecalho('2) criar_curso("Engenharia de Software", DEMA)')
        es = api.criar_curso(d_ema["id"], "Engenharia de Software")
        print(f"    -> retorno: {es['id']} ({es['nomeCurso']})")

        cabecalho('3) criar_curso("Redes de Computadores", DEMA)')
        rc = api.criar_curso(d_ema["id"], "Redes de Computadores")
        print(f"    -> retorno: {rc['id']} ({rc['nomeCurso']})")

        cabecalho("4) matricular_aluno(AlunoGraduacao Diego, ES)")
        diego = {"tipo": "AlunoGraduacao", "id": 10, "nome": "Diego Souza", "matricula": "2025010"}
        print(f"    aluno (cliente): {descrever_aluno(diego)}")
        api.matricular_aluno(es["id"], diego)

        cabecalho("5) matricular_aluno(AlunoIC Elena, ES)")
        elena = {"tipo": "AlunoIC", "id": 11, "nome": "Elena Rocha", "matricula": "2025011", "diasTrabalhados": 25, "valorBolsa": 800.0}
        print(f"    aluna (cliente): {descrever_aluno(elena)}")
        api.matricular_aluno(es["id"], elena)

        cabecalho("6) matricular_aluno(AlunoPosGraduacao Felipe, RC)")
        felipe = {"tipo": "AlunoPosGraduacao", "id": 12, "nome": "Felipe Alves", "matricula": "2025012", "diasTrabalhados": 28, "valorBolsa": 2500.0}
        print(f"    aluno (cliente): {descrever_aluno(felipe)}")
        api.matricular_aluno(rc["id"], felipe)

        cabecalho("7) emitir_folha_pagamento(Elena)")
        folha_elena = api.emitir_folha_pagamento(elena)
        print(f"    -> retorno: R$ {folha_elena['pagamento']:.2f}")

        cabecalho("8) emitir_folha_pagamento(Felipe)")
        folha_felipe = api.emitir_folha_pagamento(felipe)
        print(f"    -> retorno: R$ {folha_felipe['pagamento']:.2f}")

        cabecalho("9) listar_alunos_curso(ES)")
        for a in api.listar_alunos_curso(es["id"]):
            print(f"    - {descrever_aluno(a)}")

        cabecalho("10) listar_cursos_departamento(DEMA)")
        for c in api.listar_cursos_departamento(d_ema["id"]):
            print(f"    - {c['nomeCurso']}")

        cabecalho("11) emitir_folha_pagamento(Diego) -- caso de erro esperado")
        try:
            api.emitir_folha_pagamento(diego)
            print("    !! deveria ter falhado")
        except RuntimeError as e:
            print(f"    -> erro recebido do servidor: {e}")

        print()
        print("[cliente-python] Cenarios concluidos.")
    except Exception as e:
        print(f"[cliente-python] Falha: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: cliente-js/cliente.js
import { ApiServico } from "./apiServico.js";

function cabecalho(titulo) {
  console.log();
  console.log(`=== ${titulo} ===`);
}

function descreverAluno(a) {
  return `${a.tipo}[id=${a.id}, nome=${a.nome}, matricula=${a.matricula}]`;
}

async function main() {
  const baseUrl = process.argv[2] || process.env.SERVIDOR_URL || "http://localhost:8080";
  const api = new ApiServico(baseUrl);
  console.log(`[cliente-js] conectado em ${baseUrl}`);

  try {
    cabecalho('1) criarDepartamento("DComp - Computacao")');
    const dComp = await api.criarDepartamento("DComp - Computacao");
    console.log(`    -> retorno: ${dComp.id} (${dComp.nome})`);

    cabecalho('2) criarCurso("Ciencia da Computacao", DComp)');
    const cc = await api.criarCurso(dComp.id, "Ciencia da Computacao");
    console.log(`    -> retorno: ${cc.id} (${cc.nomeCurso})`);

    cabecalho('3) criarCurso("Sistemas de Informacao", DComp)');
    const si = await api.criarCurso(dComp.id, "Sistemas de Informacao");
    console.log(`    -> retorno: ${si.id} (${si.nomeCurso})`);

    cabecalho("4) matricularAluno(AlunoGraduacao Ana, CC)");
    const ana = { tipo: "AlunoGraduacao", id: 1, nome: "Ana Silva", matricula: "2024001" };
    console.log(`    aluno (cliente): ${descreverAluno(ana)}`);
    await api.matricularAluno(cc.id, ana);

    cabecalho("5) matricularAluno(AlunoIC Bruno, CC)");
    const bruno = { tipo: "AlunoIC", id: 2, nome: "Bruno Costa", matricula: "2024002", diasTrabalhados: 20, valorBolsa: 700.0 };
    console.log(`    aluno (cliente): ${descreverAluno(bruno)}`);
    await api.matricularAluno(cc.id, bruno);

    cabecalho("6) matricularAluno(AlunoPosGraduacao Carla, SI)");
    const carla = { tipo: "AlunoPosGraduacao", id: 3, nome: "Carla Lima", matricula: "2024003", diasTrabalhados: 30, valorBolsa: 2200.0 };
    console.log(`    aluna (cliente): ${descreverAluno(carla)}`);
    await api.matricularAluno(si.id, carla);

    cabecalho("7) emitirFolhaPagamento(Bruno)");
    const folhaBruno = await api.emitirFolhaPagamento(bruno);
    console.log(`    -> retorno: R$ ${folhaBruno.pagamento.toFixed(2)}`);

    cabecalho("8) emitirFolhaPagamento(Carla)");
    const folhaCarla = await api.emitirFolhaPagamento(carla);
    console.log(`    -> retorno: R$ ${folhaCarla.pagamento.toFixed(2)}`);

    cabecalho("9) listarAlunosCurso(CC)");
    const alunosCC = await api.listarAlunosCurso(cc.id);
    for (const a of alunosCC) {
      console.log(`    - ${descreverAluno(a)}`);
    }

    cabecalho("10) listarCursosDepartamento(DComp)");
    const cursos = await api.listarCursosDepartamento(dComp.id);
    for (const c of cursos) {
      console.log(`    - ${c.nomeCurso}`);
    }

    cabecalho("11) emitirFolhaPagamento(Ana) -- caso de erro esperado");
    try {
      await api.emitirFolhaPagamento(ana);
      console.log("    !! deveria ter falhado");
    } catch (e) {
      console.log(`    -> erro recebido do servidor: ${e.message}`);
    }

    console.log();
    console.log("[cliente-js] Cenarios concluidos.");
  } catch (e) {
    console.error("[cliente-js] Falha:", e.message);
    process.exit(1);
  }
}

main();
